### DISCUSSION TITLE
Resolve string to python object

### DISCUSSION THREAD

**Discussion by greaber** [facebookresearch/hydra#2312] [Q&A] [answered]
This seems like a dumb question, but how do I resolve a string like `torch.profiler.ProfilerActivity.CPU` to the corresponding python object?  Like instantiation, but I don't want the `_target_` to be applied to anything (because it is not even callable).  I thought about using a customer OmegaConf resolver for this, but I'm thinking probably it is already builtin to hydra, since instantiation has to do these lookups anyway, and I'm just not seeing it in the docs

**Comment by Jasha10** [comment, marked as answer]
Hi @greaber!
There are the super-useful `hydra.utils.get_class` and `hydra.utils.get_method` functions. They are undocumented currently, but that should change soon.
```python
>>> import hydra
>>> assert hydra.utils.get_method("functools.lru_cache") is __import__("functools").lru_cache
```
The `get_class` function allows looking up any class (i.e. any instance of `type`), and throws a `ValueError` if the looked-up object is not a `type`. Meanwhile, `get_method` does the same thing, except it allows looking up any `callable` object. There is a [feature request](https://github.com/facebookresearch/hydra/issues/2139) open to add a `hydra.utils.get_object` method that would do the same thing but without the restriction that the looked-up object must be a `type`/`callable`. Please give that issue a thumbs-up if such a feature would be useful to you.

I do sometimes find myself using an omegaconf resolver that calls Python's `builtins.eval`. Here's the pattern I've gravitated towards in my own projects:
```python
>>> from omegaconf import OmegaConf
>>> OmegaConf.register_new_resolver("eval", eval)
>>> yaml_data = """foo: '${eval:"bytes(123)"}'"""
>>> assert OmegaConf.create(yaml_data).foo == bytes(123)
```
To register this resolver would be unsafe if there's a chance that your DictConfig object will contain untrusted 3rd party data, as the 3rd party could inject some things to be eval'd on your machine. (The same caveat applies by-and-large to calling Hydra's `instantiate`)

**Reply by Jasha10** [reply]
Here's another thing that comes to mind:
Starting with Hydra v1.2, the `instantiate` function allows use of a [`_partial_`](https://hydra.cc/docs/advanced/instantiate_objects/overview/#partial-instantiation) keyword to wrap the given `_target_` in a `functools.partial` instance.

```
from hydra.utils import instantiate
from functools import partial, lru_cache
from omegaconf import OmegaConf

yaml_data = """
obj:
    _target_: functools.lru_cache
    _partial_: true
"""
cfg = OmegaConf.create(yaml_data)
instance = instantiate(cfg.obj)
assert isinstance(instance, partial)
assert instance.func is lru_cache
```
Incidentally, this approach does not work with `torch.profiler.ProfilerActivity.CPU` from your question above, as `instantiate` requires that the `_target_` is callable (and `torch.profiler.ProfilerActivity.CPU` is not callable).

**Reply by Jasha10** [reply]
Since, none of the options above allow instantiating `torch.profiler.ProfilerActivity.CPU`, here are two ways to do so:

#### Using the private `hydra.utils._locate` method:

```python
import torch
from hydra.utils import instantiate
from omegaconf import OmegaConf

cfg = OmegaConf.create(
    """
obj:
    _target_: hydra.utils._locate
    path: torch.profiler.ProfilerActivity.CPU
"""
)
instance = instantiate(cfg.obj)
assert instance is torch.profiler.ProfilerActivity.CPU
```
This is probably the cleanest pattern for achieving what you originally asked about, though it makes use of a private function (`_locate`).
In the future, the propsed `get_object` function (mentioned above) should cover this use case.

#### Instantiating `torch.profiler.ProfilerActivity`

```python
import torch
from hydra.utils import instantiate
from omegaconf import OmegaConf

cfg = OmegaConf.create(
    """
obj:
    _target_: torch.profiler.ProfilerActivity
    _args_:
      - 0
"""
)
instance = instantiate(cfg.obj)
assert instance == torch.profiler.ProfilerActivity.CPU
```

**Reply by greaber** [reply]
Thanks!
